### DISCUSSION TITLE
[1.1.434] IP-Adapter faceid portrait

### DISCUSSION THREAD

**Discussion by huchenlei** [Mikubill/sd-webui-controlnet#2530] [General]
## How to use
You can download the model here:
https://huggingface.co/h94/IP-Adapter-FaceID/blob/main/ip-adapter-faceid-portrait_sd15.bin
The model should be paired with `ip-adapter_face_id` preprocessor. No LoRA is required for this model.

![1705948034973](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/cdf6baa6-7a5b-4685-9671-a4ab9edb6e1f)

## Model comparisons
Following samples are generated by running
```sh
 python -m pytest -vv --junitxml=test/results.xml --cov .\extensions\sd-webui-controlnet\ --cov-report=xml --verify-base-url .\extensions\sd-webui-controlnet\tests\web_api\full_coverage\ipadapter_test.py
```
You need to set environment variable `CONTROLNET_TEST_FULL_COVERAGE` to any value in order for the test to be detected.
Some other preps are documented in https://github.com/Mikubill/sd-webui-controlnet/blob/main/tests/README.md

Checkpoint used: realisticVisionV20_v20.safetensors [e6415c4892]

### Single input
Input image
![('ip-adapter_face_id', 'ip-adapter-faceid_sd15  0a1757e9 ', 'ip-adapter-faceid_sd15_lora')_1](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/cd6a15d3-7401-442f-a9b9-7ed77c9b0192)

FaceID
![AdapterSetting(module='ip-adapter_face_id', model='ip-adapter-faceid_sd15  0a1757e9 ', lora='ip-adapter-faceid_sd15_lora')_no_neg_0](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/9570db7f-1fb6-4c29-acd5-cdcba42a3c7d)

FaceID plus
![AdapterSetting(module='ip-adapter_face_id_plus', model='ip-adapter-faceid-plus_sd15  d86a490f ', lora='ip-adapter-faceid-plus_sd15_lora')_no_neg_0](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/06007945-3f2b-4851-8907-1188582bdaf9)

FaceID plus v2
![AdapterSetting(module='ip-adapter_face_id_plus', model='ip-adapter-faceid-plusv2_sd15  6e14fc1a ', lora='ip-adapter-faceid-plusv2_sd15_lora')_no_neg_0](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/0bf0aedd-019f-436a-820c-fc7a905fcb2d)

FaceID portrait
![AdapterSetting(module='ip-adapter_face_id', model='ip-adapter-faceid-portrait_sd15  b2609049 ', lora=None)_no_neg_0](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/d163a631-6db2-42f4-91a2-bb504e3ee321)

### Multiple inputs 
- Uses 6 images under tests/images/portrait
- After #2539, you can use weight = 1.0 intead of 1/6.

FaceID
![multi_inputs_AdapterSetting(module='ip-adapter_face_id', model='ip-adapter-faceid_sd15  0a1757e9 ', lora='ip-adapter-faceid_sd15_lora')_no_neg_0](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/c029a739-7465-46fa-bae6-aecec2ca2250)

FaceID plus
![multi_inputs_AdapterSetting(module='ip-adapter_face_id_plus', model='ip-adapter-faceid-plus_sd15  d86a490f ', lora='ip-adapter-faceid-plus_sd15_lora')_no_neg_0](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/e159f18c-5cbc-4991-b4ec-9255e8c670e5)

FaceID plus v2
![multi_inputs_AdapterSetting(module='ip-adapter_face_id_plus', model='ip-adapter-faceid-plusv2_sd15  6e14fc1a ', lora='ip-adapter-faceid-plusv2_sd15_lora')_no_neg_0](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/eddcb854-923e-4e4a-93f4-4f21df5e2c97)

FaceID portrait
![multi_inputs_AdapterSetting(module='ip-adapter_face_id', model='ip-adapter-faceid-portrait_sd15  b2609049 ', lora=None)_no_neg_0](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/0e51965d-9740-40a8-aceb-d1a6fec4e46f)

### Conclusion
The new portrait model performs best when given multiple inputs.

## Notes
- Realistic vision checkpoints performs better on western faces. I used it to test whether portrait model can correctly restore features of an asian face.

## Update 2024-01-23
There was some bugs in our testing code previously. I have fixed those bugs. The model comparison samples are updated. Negative prompt has big impact on generation result according to our testing. You should avoid use a long list of negative prompt to obtain better id preservation.

**Comment by IndigoDosSantos** [comment]
I'm impressed with the new feature that allows us to view uploaded batch images directly in the controlnet. This improvement eliminates the risk of errors due to incorrect path syntax. 
While I haven't tested it thoroughly, it seems like the portrait IP-adapter might be faster than others from the faceid family. Also, its ability to blend faces effectively and maintain consistency across various prompts and seeds is quite remarkable (have a look at the image below). It effectively acts like an 'instant LoRA' as @huchenlei pointed out in [#2527 ](https://github.com/Mikubill/sd-webui-controlnet/discussions/2527), and could possibly be used to generate image sets for actual LoRAs.
![puzzle_fixed_together_transparency_psMerge3-01](https://github.com/Mikubill/sd-webui-controlnet/assets/130935112/1e14ebeb-f0e1-4e3f-98e2-07f78df2f18a)

**Comment by yamfun** [comment]
Is it normal that the old IP adaptor like "ip-adapter_clip_sd15" / "ip-adapter_sd15" no longer work after the update?

and like either the Weight or XYZ plot "[ControlNet] Weight" are not making a difference...

**Reply by huchenlei** [reply]
Can you file an issue sharing your setup and results? There might be some issues regarding the ip-adapter implementation change.

**Reply by huchenlei** [reply]
See https://github.com/Mikubill/sd-webui-controlnet/issues/2558.

**Reply by yamfun** [reply]
Updated the extension again upon seeing your reply, it works now. Thanks a lot.

**Reply by yaboosat** [reply]
@huchenlei 
Unfortunately I'm not able to see ip-adapter_face_id in my preprocessor list. I believe it's because my controlnet is old version. But when I try to update it to latest version through automatic1111 extension nothing happens :(
![s3](https://github.com/Mikubill/sd-webui-controlnet/assets/157692204/9cf737a3-d388-4114-9e3f-067b0867af79)
ns.

**Reply by IndigoDosSantos** [reply]
I'd try a fresh ControlNet install. Just make sure you don't accidentally delete the models as well, as you probably don't want to download them again.

**Comment by sipie800** [comment]
great work! it just shows obvious improvement over face id. can we have sdxl version model?

**Reply by huchenlei** [reply]
https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-portrait_sdxl.bin

**Comment by hungtooc** [comment]
I have 2 question:
1. In the original model (https://huggingface.co/h94/IP-Adapter-FaceID), only ip-adapter faceid portrait support multiple inputs. How your implement support multiple inputs for all version of ip-adapter faceid? 
2. does this extension support concat two image of people into one image like this (image from comfyUI)  
![image](https://github.com/user-attachments/assets/84c1355d-2c90-4d19-8ed3-c0e53e568b33)

4. @huchenlei
